rogue/src/components/melee.rs

```diff
@@ -32,9 +32,39 @@ impl CombatStats {
 }
 
 
+#[derive(Clone)]
+pub struct MeeleAttackRequest {
```

Comment:
Ok, I have a question about rust here.

My initial try here was 

```rust
ub struct MeeleAttackRequest {
	
Comment
    pub source: &Entity,
    pub target: &Entity,
    pub critical: bool
}
```

Which doesn't work immediately because of ownership and lifetimes. The issue is that a `MeeleAttackRequest` could outlive its contained `Entities`. I fooled around with this a bit, but didn't quite get there, I'm curious if and how this approach can work.

Switching it to an `Entity` *does* work, but I don't really understand how. The `Entity` is already contained in the ECS, which I would assume is the owner in some sense. Putting the entity itself, not a reference, into the `MeeleAttackRequest` would seem to create two owners. I don't really get why this passes the borrow checker.

Reply:
I believe switching to `Entity` works because `Entity` derives the `Copy` trait. This means when you pass an `Entity` not-by-reference Rust implicitly performs a bit-wise copy of the underlying structure. `Copy` is what's considered a "marker trait" - it doesn't actually contain any functionality itself, but it marks the type it's applied to as being efficiently `Clone`-able, and change's how the borrow checker treats the type when you get an assignment like `let b = a;`.

Normally, `let b = a;` would _move_ ownership. `let b = a.clone();` would create a copy of `a` with _new_ ownership, assuming `a` implements `Clone`. When `a` implements `Copy` the implementer is asserting to the Rust compiler that the `Clone` operation is a cheap bit-wise copy, which _changes the semantics` of `let b = a;` into a _copy assignment_.

Reply:
Oh, you also asked:
> I'm curious if and how this approach can work
You might be able to provide an explicit lifetime to essentially "force" the `MeeleAttackRequest` and entities to share a lifetime, but figuring out the "right" lifetime is challenging and probably not worth it, in my opinion. If you were really trying to optimize copy/clones at that level then it sort of suggests to me that it should be baked deeper into the underlying engine as a broader overall strategy.

Reply:
Alright, that makes sense. There must still be some trikyness going on though, because if we're really adding a *copy* of the entities into the `MeeleAttackRequest`, we *do* want to resolve to the original entities when the request is processed. I.e., when we process the requests, we deal actual damage (i.e., mutate the combat stats) on the original entity, it works correctly. We don't, for example, see that the attack request actually spawns a new goblin, that takes damage, and now there's two goblins.